WIP, RFC: Bounding sphere decimation

h3mosphere opened the pull request:
This adds LOD decimation based on bounding spheres. This works for me, I wrote it about a week ago, and I just wanted to show it, in case you find the use case intriguing.

I'm not 100% sure about the implementation, and that I got the the generic trait bounds correct.

One thing I thought of was, instead of having nested bounding spheres, have an origin, radius, initial error, and error increment per distance unit, for continuous LOD.

Anyway, thanks for the cool library, and it was pretty nice how easy it was to add this functionality, without it having necessarily being a use case you had envisioned. :+1:

h3mosphere commented:
Also, the `hide some unused warnings` commit snuck in there without me noticing. Apologies there.

dima634 reviewed (changes requested):
Left a couple of nit-pick comments. Great work! And I would love to merge it. Please let me know if you do not want to spend time fixing them so I will do it myself.

The example in the simplification section of `README.md` is outdated by these changes and needs to be updated.

h3mosphere commented on src/decimation/edge_decimation.rs:
```diff
@@ -347,3 +347,96 @@ where
+{
+    pub fn contains_point(&self, point: &Point3<TMesh::ScalarType>) -> bool {
+        nalgebra::distance(&self.origin, point) <= self.radius
+    }
```
Done. The BoundingSphere could actually be replaced by Sphere3, if I added contains_point to it?

 I only just noticed this, BoundingSphere is currently in an awkward place in the code.

I'll change this if you agree.

dima634 replied:
Agree 👍

h3mosphere commented on src/decimation/edge_decimation.rs:
```diff
@@ -181,14 +184,8 @@ where
-            None => self.max_error = TMesh::ScalarType::infinity(),
-        };
+    pub fn max_error(mut self, max_error: TMaxError) -> Self {
+        self.max_error = max_error;
```
Done. There is some slightly awkward unwrapping code as `MaxError` is not copy/clone, but it works well enough.

I should add some tests I guess to make sure it works as anticipated. I'm not sure the what the default's resolve to be, etc.

dima634 replied:
The default value should be something like `infinity` or `max`. In other words, the default should be such that condition `error < max_error` is always `true`.

h3mosphere commented:
Reworking the code, I realized that perhaps instead of MaxError, the trait could be 

```
trait DecimateEdge {
    fn should_decimate(&self, cost: ScalarType, , mesh: &TMesh, edge: &TMesh::EdgeDescriptor) -> bool
}
```

This allows for essentially NOOP for the case where we disable MaxError. i.e. we could default to IdentityDecimateEdge, which always returns false(?)

h3mosphere commented:
I also removed the assets/surface*.stl files, the source was 11MB. I'll add back in a 2d triangulation for the source plane.

dima634 commented:
> Reworking the code, I realized that perhaps instead of MaxError, the trait could be
> 
> ```
> trait DecimateEdge {
>     fn should_decimate(&self, cost: ScalarType, mesh: &TMesh, edge: &TMesh::EdgeDescriptor) -> bool
> }
> ```
> 
> This allows for essentially NOOP for the case where we disable MaxError. i.e. we could default to IdentityDecimateEdge, which always returns false(?)

Totally makes sense. I think this would be a cleaner solution.

h3mosphere commented:
OK, `MaxError` trait has been renamed to `EdgeDecimationCriteria`.

There is not currently an `IdentityDecimationCriteria`, this may make sense where None is passed as the decimation_criteria, to allow essentially a NOOP.  This however depends on the type that is set, or inferred. Currently setting the decimator to None requires the `TEdgeDecimationCriteria` to be qualified. It's a bit of a code smell, but not 100% sure what to do here. I hope I've explained this well enough. For example in  examples/simplification.rs, setting line 15 to:

`let mut decimator = EdgeDecimator::new().decimation_criteria(None);`

Causes an error.

However this works: 
`let mut decimator = EdgeDecimator::<_, ConstantErrorDecimationCriteria<_>>::new().decimation_criteria(None);`

Might need some architectural re-jig.

dima634 commented:
> OK, `MaxError` trait has been renamed to `EdgeDecimationCriteria`.
> 
> There is not currently an `IdentityDecimationCriteria`, this may make sense where None is passed as the decimation_criteria, to allow essentially a NOOP. This however depends on the type that is set, or inferred. Currently setting the decimator to None requires the `TEdgeDecimationCriteria` to be qualified. It's a bit of a code smell, but not 100% sure what to do here. I hope I've explained this well enough. For example in examples/simplification.rs, setting line 15 to:
> 
> `let mut decimator = EdgeDecimator::new().decimation_criteria(None);`
> 
> Causes an error.
> 
> However this works: `let mut decimator = EdgeDecimator::<_, ConstantErrorDecimationCriteria<_>>::new().decimation_criteria(None);`
> 
> This might need some architectural re-jig.

As an option, instead of `None` we can have `DecimateAlways` struct implementing `EdgeDecimationCriteria` trait. So instead of passing `None` you can set criteria to `DecimateAlways`. In this case, the compiler will be able to infer the generic type correctly and the code remains readable. The drawback is that users have to specify `decimation_criteria` explicitly every time but I think it is not such a big deal.
```rust
// edge_decimation.rs
pub struct IncrementalDecimator<TMesh, TCollapseStrategy, TEdgeDecimationCriteria>
where 
    TMesh: EditableMesh + TopologicalMesh + MeshMarker, 
    TCollapseStrategy: CollapseStrategy<TMesh>,
    TEdgeDecimationCriteria: EdgeDecimationCriteria<TMesh>

{
    decimation_criteria: Option<TEdgeDecimationCriteria>, // Remove `Option` here
    min_faces_count: usize,
    min_face_quality: TMesh::ScalarType,
    priority_queue: BinaryHeap<Contraction<TMesh>>,
    not_safe_collapses: Vec<Contraction<TMesh>>,
    collapse_strategy: TCollapseStrategy
}

#[derive(Debug, Default)]
pub struct DecimateAlways;

impl<TMesh: Mesh> EdgeDecimationCriteria<TMesh> for DecimateAlways {
    fn should_decimate(&self, _error: TMesh::ScalarType, _mesh: &TMesh, _edge: &TMesh::EdgeDescriptor) -> bool {
        return true;
    }
}

// simplification.rs
let mut decimator2 = EdgeDecimator::new().decimation_criteria(ConstantErrorDecimationCriteria::new(0.01f32));
decimator2.decimate(&mut mesh);
let mut decimator1 = EdgeDecimator::new().decimation_criteria(DecimateAlways);
decimator1.decimate(&mut mesh);
```

h3mosphere commented:
> As an option, instead of `None` we can have `DecimateAlways` struct implementing `EdgeDecimationCriteria` trait. So instead of passing `None` you can set criteria to `DecimateAlways`. In this case, the compiler will be able to infer the generic type correctly and the code remains readable. The drawback is that users have to specify `decimation_criteria` explicitly every time but I think it is not such a big deal.

This sounds good, I'm happy with whatever you decide is best for the library.

The behaviour may have changed a little bit with respect to how decimation_criteria and min_faces_count interact with each other, in  a default behaviour mode. But I havent used the code enough to have an opinion on that.

I'll make those changes.

h3mosphere commented:
OK, I have made those changes.

I added `AlwaysDecimate`, and `NeverDecimate` trait implementors. Changed from `DecimateAlways`, as `DecimateNever` seemed awkward.

Also updated usage comments where appropriate, and documented the new trait.

Hopefully this is now done :D

h3mosphere commented:
Little screenshot:

![Screenshot from 2023-08-28 10-56-17](https://github.com/dima634/baby_shark/assets/129932586/cfbe5573-8c03-4c48-b3b3-0963bdfe9736)

h3mosphere commented:
It would be nice to leave edges alone on non-manifold meshes. Not sure if this is possible, any thoughts appreciated.

dima634 commented:
Approved. Feel free to merge it. Thanks for working on the comments.

> It would be nice to leave edges alone on non-manifold meshes. Not sure if this is possible, any thoughts appreciated.

I think it is possible. We just need to make sure that these edges stay topologically unconnected to the opposites. Otherwise, iterators that are doing topological queries may get stuck in an infinite loop. Overall, it is probably more complicated than sounds. Right now non-manifold edges are removed [here](https://github.com/dima634/baby_shark/blob/af093bf75a27cab2201f61458ef0ab25a1580d6a/src/mesh/corner_table/table.rs#L156).

h3mosphere commented:
Not sure what you mean by merge..? I don't have commit rights..

dima634 commented:
Didn't know that. Sorry for the confusion.

dima634 commented:
Would you like to get collaborator access?
